client/network/src/protocol/block_requests.rs

```diff
@@ -163,16 +196,38 @@ where
 		BlockRequests {
 			config: cfg,
 			chain,
+			peers: HashMap::new(),
 			outgoing: FuturesUnordered::new(),
 			pending_events: VecDeque::new(),
 		}
 	}
 
 	/// Issue a new block request.
 	///
+	/// Cancels any existing request targeting the same `PeerId`.
```

Comment:
Why is it safe to invalidate requests to a peer if more requests are sent to the same peer?

Reply:
It's not a matter of safety. It's a design decision by the sync code to only allow one request per node.

Reply:
From an external point of view, previous requests are "cancelled", but internally we just forget about them. If a response comes and it doesn't match a known request, we discard that response.